packages/shared-metrics/src/util/dependencyDistances.js

```diff
@@ -0,0 +1,256 @@
+/*
+ * (c) Copyright IBM Corp. 2021
+ * (c) Copyright Instana Inc. and contributors 2021
+ */
+
+'use strict';
+
+const assert = require('assert');
+const fs = require('fs');
+const path = require('path');
+
+const CountDownLatch = require('./CountDownLatch');
+
+let logger = require('@instana/core').logger.getLogger('metrics');
+
+/**
+ * @param {import('@instana/core/src/logger').GenericLogger} _logger
+ */
+exports.setLogger = function setLogger(_logger) {
+  logger = _logger;
+};
+
+/** @type {Object.<string, any>} */
+let distancesFromRoot;
+
+/** @type {import('./CountDownLatch')} */
+let globalCountDownLatchAllPackages;
+
+/** @type {number} */
+let started;
+
+/**
+ * Calculates the distance for all dependencies, starting at the given package.json file. Direct dependencies listed in
+ * the passed package.json have distance 1. Dependencies of those dependencies have distance 2, and so on. This is
+ * calculated by parsing package.json files recursively, traversing the tree of dependencies.
+ *
+ * @param {*} packageJsonPath the path to the package.json file to examine initially
+ * @param {(distances: Object<string, any>) => void} callback
+ */
+function calculateDistancesFrom(packageJsonPath, callback) {
+  started = Date.now();
+  assert.strictEqual(typeof packageJsonPath, 'string');
+  assert.strictEqual(typeof callback, 'function');
+  distancesFromRoot = {};
```

Comment:
That's a very good point. I'll try to modify the code so that it is safe against being called multiple times concurrently. That really shouldn't happen, but it is definitely good to make it more robust.

Reply:
The suggestion to turn it into a class was spot on, it is much cleaner now.